Documentation/botr/readytorun-format.md

```diff
@@ -1,15 +1,77 @@
+* Easier deployment: desire to avoid having to deal with potentially hundreds or thousands of assemblies.
+* Side-by-side deployment: a self-contained app doesn't need to care about the location of its dependencies, their proper versions, about sharing them with other apps (including other versions of the same app).
+
+Semantic requirements for the composite R2R format flow from the combination of the basic design principle that the R2R file is a code cache to save jitting time and from the fact that the tighter coupling can be utilized to improve codegen quality.
+
+### R2R file is a code cache
+
+A corollary of this basic principle is that the composite file must maintain full functionality of single-input R2R files w.r.t. things like reflection, dynamic or generics. To achieve this goal the composite R2R file must store all the original MSIL metadata blobs for all input assemblies in the output R2R file. This is a substantial deviation from the previous format that only had one COR header addressed by the COM data directory entry in the COFF header.
+
+To achieve this goal, we propose using two complementary strategies:
+
+* In the composite R2R file, there must be a new table of COR headers and metadata blobs representing the MSIL metadata from all the input assemblies. The table must be indexable by simple assembly name for fast lookup. 
+
+* We propose slightly variating the purpose of the "main" COR header and metadata blob: in composite R2R files, where a "singleton" ECMA-335 metadata blob doesn't make much sense, we'd reuse the "main" metadata for the manifest metadata.
+
+* We may be able to leverage this semantic shift to include some virtual "composite assembly information" in the assembly attributes of the manifest metadata assembly to facilitate versioning and deployment.
+
+**Note:** As of this proposal we're not addressing managed C++ code including both native and managed code. For pre-built native code, it's generally not possible to relocate RVA's as the fixup table information is not rich enough. As a consequence, it's simply not possible to roll multiple managed C++ modules containing native code into a single output R2R PE binary. We might be able to consider some mitigations like allowing a single managed binary containing native code in a composite build but I'm not sure whether they are worth the effort and desirable in general.
+
+### Better codegen in composite R2R files
+
+The perf improvements for composite R2R files can stem from the following facts:
+
+* The composite R2R file forms a tightly-bound version bubble where an individual assembly cannot be accidentally replaced on disk so we don't need to emit any type layout checks.
```

Comment:
I think it depends on what you refer to by "messing up". For the sake of simplicity let's assume the compiler is flawless so it doesn't "mess up" just by itself. When it generates a composite file, it's a correct runnable composite file, period. It can only become invalid when tampered with by binary patching. I believe most software publishers and deployment tools address this by means of various security hashes, checksums, envelopes or digital signatures.

In the "multi-file composite R2R file" that we more or less have today - a number of assemblies compiled into the same number of R2R output files just relaxing inlining and other limits, directly embedding field offsets and the like - is much more vulnerable to replacing a given component file with a wrong file, e.g. a slightly mismatched version. The combined absence of R2R resiliency checks and any file-level consistency checks creates a hole that can cause extremely tricky bugs (e.g. slightly shifted field offsets are a big fun to debug, I know what I'm talking about).

I believe the reason why Michal is working on the PR below is the general opinion that accidental replacement of a file with a mismatched version, be it due to a bug in some deployment script, due to some weird non-deterministic condition like an antivirus holding a certain image file locked during reinstallation or some weird condition while unpacking a "single-file self-extracting package" into a temporary location, is a much higher probability and is less catered for by means ensuring consistency of individual files as mentioned above. I may be wrong and I should probably emphasize that I'm not trying to represent or interpret any official company position on this subject.

Reply:
@trylek This paragraph starts with "The perf improvements for composite R2R files can stem from the following facts", but the points that follow are more about deployment robustness - not about performance. I think it is really what @mjsabby is getting at.

Reply:
Upon rereading I agree the section (which now resides in the design file) is somewhat misleading in multiple aspects - the type layout checks are omitted even in the multi-file large version bubble case, the MVID checks probably have minimum perf effect and more or less the only meaningful argument is the generics but even there the actual perf benefit is hard to demonstrate. I guess I should reword this to more focus on actual code optimizations like embedding field offsets or future support for direct calls.